clickhouse-admin/src/context.rs

```diff
@@ -3,84 +3,555 @@
 use slog::Logger;
+use slog::{error, info};
+use std::fs::File;
+use std::io::{BufRead, BufReader};
 use std::net::SocketAddrV6;
-use std::sync::Arc;
-use tokio::sync::Mutex;
+use std::str::FromStr;
+use tokio::sync::{oneshot, watch};
 
 pub struct KeeperServerContext {
     clickward: Clickward,
     clickhouse_cli: ClickhouseCli,
     log: Logger,
+    pub generate_config_tx: Sender<GenerateConfigRequest>,
+    pub generation_rx: watch::Receiver<Option<Generation>>,
 }
 
 impl KeeperServerContext {
-    pub fn new(clickhouse_cli: ClickhouseCli) -> Self {
-        let log = clickhouse_cli
-            .log
-            .new(slog::o!("component" => "KeeperServerContext"));
+    pub fn new(
+        log: &Logger,
+        binary_path: Utf8PathBuf,
+        listen_address: SocketAddrV6,
+    ) -> Result<Self> {
+        let clickhouse_cli =
+            ClickhouseCli::new(binary_path, listen_address, log);
+        let log = log.new(slog::o!("component" => "KeeperServerContext"));
         let clickward = Clickward::new();
-        Self { clickward, clickhouse_cli, log }
+        let config_path = Utf8PathBuf::from_str(CLICKHOUSE_KEEPER_CONFIG_DIR)
+            .unwrap()
+            .join(CLICKHOUSE_KEEPER_CONFIG_FILE);
+
+        // If there is already a configuration file with a generation number we'll
+        // use that. Otherwise, we set the generation number to None.
+        let gen = read_generation_from_file(config_path)?;
+        let (generation_tx, generation_rx) = watch::channel(gen);
+
+        // We only want to handle one in flight request at a time. Reconfigurator execution will retry
+        // again later anyway. We use flume bounded channels with a size of 0 to act as a rendezvous channel.
+        let (generate_config_tx, generate_config_rx) = flume::bounded(0);
+        tokio::spawn(long_running_generate_config_task(
+            generate_config_rx,
+            generation_tx,
+        ));
+
+        Ok(Self {
+            clickward,
+            clickhouse_cli,
+            log,
+            generate_config_tx,
+            generation_rx,
+        })
+    }
+
+    pub async fn send_generate_config_and_enable_svc(
+        &self,
+        keeper_settings: KeeperConfigurableSettings,
+    ) -> Result<GenerateConfigResult, HttpError> {
+        let clickward = self.clickward();
+        let log = self.log();
+        let node_settings = NodeSettings::Keeper {
+            settings: keeper_settings,
+            fmri: "svc:/oxide/clickhouse_keeper:default".to_string(),
+        };
+
+        let (response_tx, response_rx) = oneshot::channel();
+        self.generate_config_tx
+            .try_send(GenerateConfigRequest::GenerateConfig {
+                clickward,
+                log,
+                node_settings,
+                response: response_tx,
+            })
+            .map_err(|e| {
+                HttpError::for_internal_error(format!(
+                    "failure to send request: {e}"
```

Comment:
I'm not sure including the `{e}` here will be useful. Maybe instead we should match on it and return different kinds of HTTP errors for the two cases? Something like

```rust
            .map_err(|e| match e {
                TrySendError::Full(_) => HttpError::for_unavail(
                    None,
                    "channel full: another config request is still running"
                        .to_string(),
                ),
                TrySendError::Disconnected(_) => {
                    HttpError::for_internal_error(
                        "long-running generate-config task died".to_string(),
                    )
                }
            })?;
```